File: LLMgeneratedCode/DeepseekR1Distill-3.py
import pandas as pd
import numpy as np
from sklearn.linear_model import Lasso
from sklearn.model_selection import GridSearchCV, train_test_split
from sklearn.metrics import mean_squared_error
import matplotlib.pyplot as plt

# Load feature data and metadata for training set
def load_data_csv(csv_path):
    # Read the .csv file with features as rows and samples as columns
    df = pd.read_csv(csv_path, header=0).T
    return df

df_training_features = load_data_csv("df_training.csv")
ano_training = pd.read_csv("ano_training.csv")

# Merge feature data with metadata using Sample_ID
training_df = pd.merge(df_training_features, ano_training, on='Sample_ID')

# Split training data into features and target variable
X_train = training_df.drop(columns=['Gestational_Age'])
y_train = training_df['Gestational_Age']

# Load test set feature data and metadata
df_test_features = load_data_csv("df_test.csv")
ano_test = pd.read_csv("ano_test.csv")

# Merge test data with metadata using Sample_ID
test_df = pd.merge(df_test_features, ano_test, on='Sample_ID')

# Split test data into features and target variable (for prediction)
X_test = test_df.drop(columns=['Gestational_Age'])
y_true = test_df['Gestational_Age']

# Perform feature scaling if necessary
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Split training data into train and validation sets for hyperparameter tuning
X_trainCV, X_val, y_trainCV, y_val = train_test_split(
    X_train_scaled, y_train, test_size=0.2, random_state=42)

# Define parameter grid for Lasso regression
param_grid = {'alpha': [0.1, 1.0, 5.0, 10.0]}

# Perform grid search with cross-validation to find optimal alpha
lasso = Lasso()
grid_search = GridSearchCV(lasso, param_grid, cv=5, scoring='neg_mean_squared_error')
grid_search.fit(X_trainCV, y_trainCV)

best_alpha = grid_search.best_params_['alpha']
print(f"Best alpha found: {best_alpha}")

# Fit final model with optimal alpha
final_lasso = Lasso(alpha=best_alpha)
final_lasso.fit(X_train_scaled, y_train)

# Make predictions on test set
y_pred = final_lasso.predict(X_test_scaled)

# Calculate RMSE
rmse = np.sqrt(mean_squared_error(y_true, y_pred))
print(f"RMSE: {rmse:.4f}")

# Create scatter plot of predicted vs actual values
plt.figure(figsize=(6, 6))
plt.scatter(y_true, y_pred, alpha=0.5)
plt.xlabel('Actual Gestational Age')
plt.ylabel('Predicted Gestational Age')
plt.title('Predicted vs Actual Gestational Age')
plt.plot([min(y_true), max(y_true)], [min(y_true), max(y_true)], 
         color='red', linestyle='--', linewidth=2, label='Perfect Prediction')
plt.text(0.9 * min(y_true), 0.95 * max(y_true), f'RMSE = {rmse:.4f}',
        verticalalignment='top', horizontalalignment='right')
plt.legend()
plt.savefig("./out2/DeepseekR1Distill.py.pdf", format="pdf")

File: LLMgeneratedCode/DeepseekR1Distill-3.R
# Load required libraries
library(dplyr)
library(ggplot2)
library(readr)

# Read training data
training_data <- read_csv("df_training.csv", rowNameKey = 1)

# Read test data
test_data <- read_csv("df_test.csv", rowNameKey = 1)

# Read metadata for sample IDs
ano_training <- read_csv("ano_training.csv")
ano_test <- read_csv("ano_test.csv")

# Extract sample IDs from training and test metadata
train_ids <- ano_training[sample_id != "ID"] %>% distinct(sample_id)
test_ids <- ano_test[sample_id != "ID"] %>% distinct(sample_id)

# Filter training data based on sample IDs
training_data <- training_data %>% filter(row == training_data[rowNameKey] %>% 
                                             group_by(sample_id) %>% 
                                             filter(sample_id %in% train_ids))

# Filter test data based on sample IDs
test_data <- test_data %>% filter(row =="testing_data[rowNameKey]" %>% 
                                     group_by(sample_id) %>% 
                                     filter(sample_id %in% test_ids))

# Remove the 'row' column from both datasets
training_data <- training_data %>% select(-row)
test_data <- test_data %>% select(-row)

# Handle missing values (remove rows with NAs)
training_data <- na.omit(training_data)
test_data <- na.omit(test_data)

# Select top 100 most variable features
set.seed(123)  # For reproducibility
feature_variance <- training_data %>% 
    summarise(variance = var(GA)) %>% 
    mutate(rowname = rowname,
           feature_name = set_rowname(.))
top_features <- feature_variance %>% 
    arrange(desc(variance)) %>% 
    slice(1:100) %>% 
    pull(feature_name)

# Preprocess the data (centering and scaling)
training_data_processed <- training_data %>% 
    select(-GA) %>% 
    mutate_all(funs((. - mean(.)) / sd(.)))

test_data_processed <- test_data %>% 
    select(-GA) %>% 
    mutate_all(funs((. - mean(.)) / sd(.)))

# Fit the model using elastic net regression (cv.glmnet)
library(glmnet)

# Prepare data for modeling
x_train <- as.matrix(training_data_processed)
y_train <- training_data$GA

set.seed(42)  # For reproducibility
cv_fit <- cv.glmnet(x = x_train, y = y_train,
                    alpha = 0.5, nlambda = 100)

# Get optimal lambda value
best_lambda <- cv_fit$lambda.min

# Make predictions on test set
x_test <- as.matrix(test_data_processed)
y_pred <- predict(cv_fit, s = best_lambda, newx = x_test)

# Calculate RMSE
rmse <- sqrt(mean((y_pred - test_data$GA)^2))
print(paste("Root Mean Squared Error (RMSE):", rmse))

# Create scatter plot of predicted vs actual GA
test_data <- test_data %>% 
    mutate(predicted_GA = y_pred) %>% 
    select(GA, predicted_GA)

ggplot(test_data, aes(x = predicted_GA, y = GA)) +
    geom_point(alpha = 0.6) +
    geom_abline(intercept = 0, slope = 1, color = "red", linetype = "dashed") +
    labs(title = paste("Predicted vs Actual Gestational Age (RMSE =", round(rmse, 3), ")"),
         x = "Predicted Gestational Age",
         y = "Actual Gestational Age") +
    theme_minimal()
